draft-ietf-httpbis-message-signatures.md

```diff
@@ -1809,6 +1865,8 @@ From here, the signing process proceeds as usual.
 
 Upon verification, it is important that the verifier validate not only the signature but also the value of the Content-Digest field itself against the actual received content. Unless the verifier performs this step, it would be possible for an attacker to substitute the message content but leave the Content-Digest field value untouched to pass the signature. Since only the field value is covered by the signature directly, checking only the signature is not sufficient protection against such a substitution attack.
 
+As discussed in {{DIGEST}}, the value of the Content-Digest field is dependent on the content encoding of the message. If an intermediary changes the content encoding, the resulting Content-Digest value would change, which would in turn invalidate the signature. Any intermediary performing such an action would need to apply a new signature with the updated Content-Digest field value, similar to the reverse proxy use case discussed in {{signature-multiple}}.
```

Comment:
@jricher Here we are implying that not signing content-encoding means that there is no content coding. @reschke do you think it is a reasonable assertion? In general, this could be a slippery slope...

```suggestion
As discussed in {{DIGEST}}, the value of the Content-Digest field is dependent on the content encoding of the message. 
The above exchange implies that when the Content-Encoding is not included in the signature, then no content coding is applied.
If an intermediary changes the content encoding, the resulting Content-Digest value would change, which would in turn invalidate the signature. Any intermediary performing such an action would need to apply a new signature with the updated Content-Digest field value, similar to the reverse proxy use case discussed in {{signature-multiple}}.
```

Reply:
Not sure what the question is.

If Content-Digest is signed, and somebody changes the content by changing the content encoding, the signature will be invalid, right? Isn't that what we want here?

Reply:
No, that's not the implication -- if Content-Encoding is not included in the signature, it doesn't mean that it's not applied, it only means that the encoding type is not signed. It could still be applied to the message if the field is there. And if it gets changed, the Content-Digest also gets changed, which fails the signature. This is desired behavior but something we want to warn implementors about, which is what the existing text does.

Reply:
In the above example we have that:

1. content-digest is signed
2. content-encoding is not present
3. content-encoding is not signed

If I add a content-encoding header, the signature is still valid, since content-encoding is not signed and the client does not know whether:

1. content-encoding in the original request was present but not signed by accident
2. content-encoding was missing and the request was tampered

Reply:
Correct, you cannot sign an absent header. That feature was discussed at length and rejected by the wg. The original message could have a content encoding header, whether or not that's covered by the signature. Absence from the signature does not imply absence from the message, as you claim above. These are independent.

If you add a content encoding header, and this changes the content, the signature is not valid because the digest is no longer valid - while the signature itself would pass the digest would not, as is pointed out in this section. Now this is assuming you change the content and don't just add a meaningless header. If you don't change the content but claim a different encoding, this should fail when the message is processed because the encoding makes no sense. Signature and digest won't fix a badly formed http message.

In all of these cases, I believe the system is behaving as intended and failing where we'd expect it to fail, and how we expect it to fail. So, I am with @reschke here in that I don't see what the issue is with this situation.

Reply:
I'm with Julian and Justin here. Not signing a field is tangential to that field applying to the message.

Signatures doesn't support covering the absence of a field. And it's sad that there's no formal definition that allows "Content-encoding: nothing". But these are things we are stuck with. The rough edge, of signing content-digest and having someone screw up the message isn't nice but will fail closed and will be quickly noticeable. The workaround is gzip (or whatever) the content if your application really can't live this sharp edge, then you have a proper content-encoding to send.

 I don't think Signstures needs to do anything special here. There's lots of ways applications can screw up signing unrelated to content digest.

Reply:
I agree that the issue is we can't say `content-encoding: nothing`. The example could use a content-encoding as a workaround.